sunpy/net/helioviewer.py

```diff
@@ -29,7 +29,43 @@ def __init__(self, url="https://api.helioviewer.org/"):
             version 2 of the API.
         """
         self._api = url
+        self._source_dict = dict()
+        self.init_src_dict()
+    
+    def init_src_dict(self):
+        datasources = self.get_data_sources() # doctest: +REMOTE_DATA
+
+        for name, observ in datasources.items():  # doctest: +REMOTE_DATA
+            if name == "TRACE":  # doctest: +REMOTE_DATA
+                for instr, params in observ.items():  # doctest: +REMOTE_DATA
+                    self._source_dict[(name, instr, 'None', 'None')] = params['sourceId']  # doctest: +REMOTE_DATA
+            else:  # doctest: +REMOTE_DATA
+                for inst, detect in observ.items():  # doctest: +REMOTE_DATA
+                    for wavelength, params in detect.items():  # doctest: +REMOTE_DATA
+                        if name in ["Hinode", "STEREO_A", "STEREO_B"] or wavelength in ["C2", "C3"]:  # doctest: +REMOTE_DATA
+                            for wave, adict in params.items():  # doctest: +REMOTE_DATA
+                                self._source_dict[(name, inst, wavelength, wave)] = adict['sourceId']  # doctest: +REMOTE_DATA
+                        else:  # doctest: +REMOTE_DATA
+                            self._source_dict[(name, inst, 'None', wavelength)] = params['sourceId']  # doctest: +REMOTE_DATA
 
+    def get_source_dict(self):
+        source_dict = dict()
+        datasources = self.get_data_sources() # doctest: +REMOTE_DATA
+
+        for name, observ in datasources.items():  # doctest: +REMOTE_DATA
+            if name == "TRACE":  # doctest: +REMOTE_DATA
+                for instr, params in observ.items():  # doctest: +REMOTE_DATA
+                    source_dict[(name, instr, 'None', 'None')] = params['sourceId']  # doctest: +REMOTE_DATA
+            else:  # doctest: +REMOTE_DATA
+                for inst, detect in observ.items():  # doctest: +REMOTE_DATA
+                    for wavelength, params in detect.items():  # doctest: +REMOTE_DATA
+                        if name in ["Hinode", "STEREO_A", "STEREO_B"] or wavelength in ["C2", "C3"]:  # doctest: +REMOTE_DATA
+                            for wave, adict in params.items():  # doctest: +REMOTE_DATA
+                                source_dict[(name, inst, wavelength, wave)] = adict['sourceId']  # doctest: +REMOTE_DATA
+                        else:  # doctest: +REMOTE_DATA
+                            source_dict[(name, inst, 'None', wavelength)] = params['sourceId']  # doctest: +REMOTE_DATA
+        return source_dict
+        
```

Comment:
Sorry but can you explain all of this for me?

It looks like you've duplicated code and moved them out of the other functions?

Reply:
Because the init_src_dict won't allow me to use both the ways defined in download_jp2 to download image. So I had to create another function for this.

Reply:
I don't think you need any of this to make it work. I just personally don't follow.

Reply:
We can download the image with the following two lines of code
        >>> filepath = hv.download_jp2('2012/07/03 14:30:00', observatory='SDO', instrument='HMI', detector='None', measurement='continuum')   # doctest: +REMOTE_DATA

and

        >>> file = hv.download_jp2('2012/07/03 14:30:00', sourceId=data_sources[('SOHO' 'LASCO' 'C2' 'white-light')]   # doctest: +REMOTE_DATA

To use the latter one we need a dictionary containing the source ID's. So I created a dictionary which can be accessed by the function

Reply:
Sure, but the second one we don't plan to advertise to the user. We don't want them to use the sourceId keyword.

Reply:
But we allowed them to use this in the older version of API so I thought to keep it. If you want we can remove that function then we won't be needing the duplicate code

Reply:
We want to keep the older API but I don't see why you need two functions.

Reply:
Also both `init_src_dict` and `get_source_dict` just seem to be the same code twice. You should also remove the comments.

Reply:
>>> file = hv.download_jp2('2012/07/03 14:30:00', sourceId=data_sources[('SOHO' 'LASCO' 'C2' 'white-light')]   # doctest: +REMOTE_DATA

Do we want to keep this or we should leave it?

Reply:
Remove it.